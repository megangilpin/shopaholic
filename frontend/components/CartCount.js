import React from 'react';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import styled from 'styled-components';

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  /* Initial State of the entered Dot */
  /* flipping it on its back */
  .count-enter {
    transform: rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: rotateX(0.5turn);
  }
`;

const Dot = styled.div`
  background-color: ${props => props.theme.lightgrey};
  border-radius: 50%;
  padding: 0.5rem;
  min-width: 3rem;
  margin-left: 1rem;
  line-height: 2.2rem;
  text-shadow: 1px 1px 1px rgba(255, 255, 255, 0.66);
  color: ${props => props.theme.medPurple};
  font-size: .9em;
  font-weight: 600;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;
//font font-feature-settings and font-variant-numeric give the same width to the number no matter the size of the number ie: '1' and '2 is printed on the screen as the same width

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 4000, exit: 4000 }}
      >
        <Dot className="cart-count">{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

export default CartCount;