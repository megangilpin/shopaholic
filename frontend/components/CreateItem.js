import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from './styles/Form';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';

class CreateItem extends Component {
  state = {
    title: 'Shoes',
    description: 'I love those shoes', 
    image: 'dog.jpg',
    largeImage: 'largeDog.jpg',
    price: 0, 
  };
 
  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value; 
    this.setState({ [name]: e.target.value })
  }

  submit = e => {
    e.preventDefault();
    console.log(this.state)
  }

  render() {
    return (
      <Form onSubmit={this.submit}>
        <fieldset>
          <label htmlFor="title">
            Title
            <input 
              type="text" 
              id="title" 
              name="title" 
              placeholder="Title" 
              required value={this.state.title}
              onChange={this.handleChange} 
            />
          </label>

          <label htmlFor="title">
            Price
            <input
              type="number"
              id="price"
              name="price"
              placeholder="Price"
              required value={this.state.price}
              onChange={this.handleChange} 
            />
          </label>

          <label htmlFor="title">
            Description
            <textarea
              id="description"
              name="description"
              placeholder="Enter a description"
              required value={this.state.description}
              onChange={this.handleChange} 
            />
          </label>
          <button type="submit">Submit</button>
        </fieldset>
      </Form>
    );
  }
}

export default CreateItem;