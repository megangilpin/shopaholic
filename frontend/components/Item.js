import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import ItemStyles from './styles/ItemStyles';
import formatMoney from '../lib/formatMoney';
import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';


class Item extends Component {
  
  render() {
    const { item } = this.props;
    return (
      <ItemStyles>
        <div className="card">
          {item.image && <img src={item.image} alt={item.title} />}
          <Link
            href={{
              pathname: '/item',
              query: { id: item.id },
            }}
            >
            <a className="card_title">{item.title}</a>
          </Link>
          <p className="card_money">{formatMoney(item.price)}</p>
          <p className="card_description">{item.description}</p>
          
          <div className="buttonList">
          
            <Link
              href={{
                pathname: 'update',
                query: { id: item.id },
              }}
            >
              <a>Edit ✏️</a>
            </Link>
            <AddToCart id={item.id} />
            <DeleteItem id={item.id}>Delete Item</DeleteItem>
          </div>
        </div>
      </ItemStyles>
    );
  }
}

Item.propTypes = {
  item: PropTypes.object.isRequired,
};

export default Item;
