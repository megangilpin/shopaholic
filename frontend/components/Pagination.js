import React from 'react';
import PaginationStyles from './styles/PaginationStyles';

const Pagination = props => (
  <PaginationStyles>
    <p>Hi i'm the Pagination!</p>
  </PaginationStyles>
)

export default Pagination;