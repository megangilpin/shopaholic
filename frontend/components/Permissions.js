import { Query, Mutation } from 'react-apollo';
import Error from './ErrorMessage';
import gql from 'graphql-tag';
import Table from './styles/Table';
import SickButton from './styles/SickButton';
import PropTypes from 'prop-types';

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
];

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;

const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`;

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => (
      <div>
        <Error error={error} />
        <div>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map(permision => 
                  <th key={permision}>{permision}</th>)}
                <th>👇🏻</th>
              </tr>
            </thead>
            <tbody>{data.users.map(user =>  <UserPermissions user={user} key={user.id} />)}</tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

class UserPermissions extends React.Component {
  // Typically you might use PropTypes.object.isRequired.  
  // Will be shaping the data since we know it will always have the info. (can't create a user without that info)
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired,
  };
  // Seeding the state = ok to use props in the state because if any updates to the date occur it comes from with in this function
  state = {
    permissions: this.props.user.permissions,
  };

  handlePermissionChange = (e, updatePermissions) => {
    const checkbox = e.target;
    console.log(e.target.value)
  // make a copy for the permission array
  let updatedPermissions = [...this.state.permissions]
  // // check if it is checked or not
    if(checkbox.checked) {
      updatedPermissions.push(checkbox.value)
    } else {
      updatedPermissions = updatedPermissions.filter(permission => permission !== checkbox.value);
    }
    // use a call back to run updatePermissions in order to ensure it will run after the state is set
    this.setState({ permissions: updatedPermissions }, updatePermissions);
    console.log(updatedPermissions)
  }
  render() {
    const user = this.props.user;
    return (
      <Mutation mutation={UPDATE_PERMISSIONS_MUTATION} 
      variables={{
        permissions: this.state.permissions,
        userId: this.props.user.id
      }} >
        {(updatePermissions, { loading, error }) => (
          <>
            {error && <tr><td colspan="8"><Error error={error} /></td></tr>}
            < tr >
              <td>{user.name}</td>
              <td>{user.email}</td>
              {possiblePermissions.map(permission => (
                <td key={permission}>
                  <label htmlFor={`${user.id}-permission-${permission}`}>
                    <input
                      id={`${user.id}-permission-${permission}`}
                      type="checkbox"
                      checked={this.state.permissions.includes(permission)}
                      value={permission}
                      onChange={(e) => 
                        this.handlePermissionChange(e, updatePermissions)
                      }
                    />
                  </label>
                </td>
              ))}
              <td>
                <SickButton type="button" disabled={loading} onClick={updatePermissions}>
                  Updat{loading ? 'ing' : 'e'}
                </SickButton>
              </td>
            </tr>
          </>
        )
        }
      </Mutation>
    );
  }
}

export default Permissions;