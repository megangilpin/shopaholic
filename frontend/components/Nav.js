import Link from 'next/link';
import { useRouter } from 'next/router'
import { Mutation } from 'react-apollo';
import { TOGGLE_CART_MUTATION } from './Cart';
import NavStyles from './styles/NavStyles';
import User from './User';
import CartCount from './CartCount';
import SignOut from './SignOut';


function Nav() {
  return (
  <User>
    {({ data }) => {
      const me = data ? data.me : null
      return (
      <NavStyles data-test="nav">
        <Link href="/items">
          <a >Shop</a>
        </Link>
        {me && (
          <>
            <Link href="/sell">
                <a >Sell</a>
            </Link>
            <Link href="/orders">
              <a >Orders</a>
            </Link>
            {/* <Link href="/me">
              <a>Account</a>
            </Link> */}
            <SignOut />
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {(toggleCart) => (
                <button className="nav-btn" onClick={toggleCart}>
                  <span className="nav-btn-text">
                    My Cart
                  </span>
                  <span className="cart-count">
                    <CartCount count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)}></CartCount>
                  </span>
                </button>
              )}
            </Mutation>
          </>
        )}
        {!me && (
          <Link href="/signup">
            <a>Sign In</a>
          </Link>

        )}
      </NavStyles>
    )
    }}
  </User>
)};

export default Nav;