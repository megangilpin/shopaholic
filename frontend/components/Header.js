import styled from 'styled-components';
import Router from 'next/router';
import NProgress from 'nprogress'
import Nav from './Nav';
import Link from 'next/link';
import Cart from './Cart';
import Search from './Search';

Router.onRouteChangeStart = () => {
  NProgress.start();
};
Router.onRouteChangeComplete = () => {
  NProgress.done();
};
Router.onRouteChangeError = () => {
  NProgress.done(); 
};

const Logo = styled.h1`
  font-size: 4rem;
  margin-left: 2rem;
  position: relative;
  z-index: 2;
  a {
    padding: 2rem;
    background: ${props => props.theme.offWhite};
    color: ${props => props.theme.lightgrey};
    text-shadow: ${props => props.theme.lgemboss};
    font-size: 7rem;
    font-weight: 400;
    letter-spacing: 0.1em;
    border: 0;
    border-radius: 10px;
    text-transform: uppercase;
    text-decoration: none;
    transition: all .3s;
    @media (max-width: 700px) {
      font-size: 5rem;
    }
  }
  a:hover {
    color:  ${props => props.theme.lightPurple};
    box-shadow: -6px -6px 10px white,
    1px 1px 5px rgba(0, 0, 0, 0.1);
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
  @media (max-width: 1300px) {
    margin: 0;
    text-align: center;
    a {
      font-size: 8rem;
      padding: 0.5rem 1rem;
    }
  }
`;

const StyledHeader = styled.header`
  .bar {
    // border-bottom: 10px solid ${props => props.theme.black};
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
    @media (max-width: 1300px) {
      grid-template-columns: 1fr;
      justify-content: center;
    }
  }
  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    // border-bottom: 1px solid ${props => props.theme.lightPurple};
  }
  hr.navbar {
    height: 10px;
    border: 0;
    background-color: ${props => props.theme.offWhite};
    box-shadow: ${props => props.theme.lgemboss};
  }
`;
const Header = () => (
  <StyledHeader>
    <div className="bar">
      <Logo>
        <Link href="/">
          <a>shopaholic</a>
        </Link>
      </Logo>
      <Nav />
    </div>
    {/* <hr className="navbar"></hr> */}
    <div className="sub-bar">
      <Search></Search>
    </div>
    <Cart></Cart>
  </StyledHeader>
)

export default Header;