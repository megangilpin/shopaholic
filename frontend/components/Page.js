import React, { Component } from 'react';
import styled, { ThemeProvider, injectGlobal } from 'styled-components';
import Header from '../components/Header'
import Meta from '../components/Meta'

const theme = {
  red: '#99991A',
  darkPurple: '#2F2E3D',
  black: '#545066',
  grey: '#726F93',
  medPurple: '#AAA9BE',
  lightPurple: '#C7C6D4',
  darkGrey: '#C0BABA',
  lightgrey: '#E3E2E9',
  offWhite: '#F6F6F6',
  maxWidth: '1000px',
  bs: '0 12px 24px 0 rgba(0, 0, 0, 0,09)',
  lgemboss: '-2px -2px 2px rgba(255, 255, 255), 1px 1px 2px #C0BABA',
  greyemboss: '-1px -2px 2px rgba(255, 255, 255) , 1px 1px 2px #545066',
  purpleEmboss: '-1px -2px 2px #E3E2E9 , 1px 1px 2px #726F93',
};

const StyledPage = styled.div`
  background:${props => props.theme.offWhite};
  color: ${props => props.theme.darkPurple};
`;

const Inner = styled.div `
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
  padding: 2rem;
`;

injectGlobal `
  html { 
    font-family: "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace;
    font-style: normal; 
    font-variant: normal; 
    box-sizing: border-box;
    font-size: 10px;
  }
  *, *:before, *:after {
    box-sizing: inherit;
  }
  body {
    font-family: "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace;
    padding:0;
    margin: 0;
    font-size: 1.4rem;
    line-height: 2;
  }
  p {
    font-size: 1.4rem; 
    font-weight: 400;
    color: ${theme.grey};
  }
  a {
    font-size: 1.4rem; 
    font-weight: 400;
    text-decoration: none;
    color: ${theme.darkGrey};
  }
  button {
    font-family: "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace;
    font-weight: 700;
    font-size: 2rem;
  }
  h3 { 
    color: ${theme.grey};
  }
  .blockquote { 
    font-family: "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace; 
    font-size: 2rem; 
    font-weight: 500; 
    line-height: 3rem; 
    color: ${theme.grey};
  }
  .smallquote { 
    font-family: "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace; 
    font-size: 1.6rem; 
    font-weight: 700; 
  }
`

class Page extends Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <StyledPage> 
          <Meta />
          <Header />
          <Inner>
            {this.props.children}
          </Inner>
        </StyledPage>
      </ThemeProvider>
    );
  }
}

export default Page;