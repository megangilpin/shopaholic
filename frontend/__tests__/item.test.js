import ItemComponent from '../components/Item';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { CURRENT_USER_QUERY } from '../components/User';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import wait from 'waait';

const fakeItem = {
  id: 'ABC123',
  title: 'T-shirt',
  price: 4000,
  description: 'Tye Dye Tshirt',
  image: 'dog.jpg',
  largeImage: 'largedog.jpg',
};

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
];

describe('<Item/>', () => {
  it('renders and matches the snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <ItemComponent item= {fakeItem}/>)
      </MockedProvider>
    );
    await wait();
    wrapper.update();

    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
  })
  // it('renders the image properly', async () => {
  //   const wrapper = shallow(<ItemComponent item={fakeItem}/>);
  //   const image = wrapper.find('img');
  //   // expect(image.props().src).toBe(fakeItem.image);
  //   expect(image.props().alt).toBe(fakeItem.title);
  // });

  // it('renders price tag and title', () => {
  //   const wrapper = shallow(<ItemComponent item={fakeItem}/>);
  //   const PriceTag = wrapper.find('PriceTag');
  //   // .dive() and .children()n will allow you to shallow render one level deeper
  //   // console.log(PriceTag.dive().text());
  //   expect(PriceTag.children().text()).toBe('$50')
  //   expect(wrapper.find('Title a').text()).toBe(fakeItem.title);
  // });

  // it('renders out the buttons properly', () => {
  //   const wrapper = shallow(<ItemComponent item={fakeItem}/>);
  //   const ButtonList = wrapper.find('.buttonList');
  //   expect(ButtonList.children()).toHaveLength(3);
  //   expect(ButtonList.find('Link')).toHaveLength(1);
  //   expect(ButtonList.find('AddToCart').exists()).toBe(true);
  //   expect(ButtonList.find('DeleteItem').exists()).toBe(true);
  // });
})