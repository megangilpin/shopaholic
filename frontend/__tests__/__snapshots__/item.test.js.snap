// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item/> renders and matches the snapshot 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Edit ✏️
  </button>,
  <button
    disabled={false}
    onClick={[Function]}
  >
    Add
     To Cart 🛒
  </button>,
  <button
    onClick={[Function]}
  >
    Delete Item
  </button>,
]
`;
